test/boost/reader_concurrency_semaphore_test.cc

```diff
@@ -1077,3 +1084,352 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_set_resources) {
+        const auto num_reads = 64;
+
+        auto num_sstables = 0;
+        parallel_for_each(boost::irange(0, sstable_write_concurrency), [&] (int i) {
+            return seastar::async([&, i] {
+                while (num_sstables != target_num_sstables) {
+                    ++num_sstables;
+                    auto sst = sst_man.make_sstable(s, sstables_dir.path().string(), sstables::generation_type{num_sstables}, sst_man.get_highest_supported_format(), sstables::sstable_format_types::big);
+                    auto writer_cfg = sst_man.configure_writer("test");
+                    sst->write_components(
+                        make_flat_mutation_reader_from_mutations_v2(s, semaphore.make_tracking_only_permit(s.get(), "test", db::no_timeout), mut, s->full_slice()),
+                        1,
+                        s,
+                        writer_cfg,
+                        encoding_stats{}).get();
+                    sst->load().get();
+                    tbl.add_sstable_and_update_cache(std::move(sst)).get();
+                }
+            });
+        }).get();
+
+        auto read_id = env.prepare("SELECT value FROM ks.tbl WHERE pk = ? AND ck = ?").get0();
+
+        // We first check that the test params are not too strict and a single
+        // read can finish successfully.
+        {
+            auto msg = env.execute_prepared(read_id, {cql3::raw_value::make_value(serialized(0)), cql3::raw_value::make_value(serialized(0))}).get();
+            assert_that(msg).is_rows().with_rows_ignore_order({ {serialized(value)} });
+        }
+
+        uint64_t successful_reads = 0;
+        uint64_t failed_reads = 0;
+
+        parallel_for_each(boost::irange(0, num_reads), [&] (int i) {
+            return env.execute_prepared(read_id, {cql3::raw_value::make_value(serialized(0)), cql3::raw_value::make_value(serialized(0))}).then_wrapped([&] (future<shared_ptr<cql_transport::messages::result_message>> fut) {
+                if (fut.failed()) {
+                    // We expect failed, OOM-killed reads here.
+                    // No way to verify why they failed so we swallow all failures.
+                    fut.ignore_ready_future();
+                    ++failed_reads;
+                    return;
+                }
+                assert_that(fut.get()).is_rows().with_rows_ignore_order({ {serialized(value)} });
+                ++successful_reads;
+            });
+        }).get();
+
+        testlog.info("total reads: {} ({} successful, {} failed)", num_reads, successful_reads, failed_reads);
+        dump_stats(semaphore, log_level::info, false);
+
+        // There should be both successful and failed reads.
+        // If there is only one or the other, the test is not testing anything.
+        // We also check that the memory limiting mechanism of the semaphore was engaged.
+        // The test is meaningless without it.
+        // In the slow debug builds we never reach the kill limit for some reason.
```

Comment:
The disk is probably faster than the asan CPU. So r-c-s executes in CPU bound mode.

We could fix that with a virtualized backing file that intentionally delays the response, so we have this pattern:

1. read 1 launches N sstable reads
2. reads are artificially held waiting
3. r-c-s has no choice but to admit more reads, the same happens to them. Now we have N*M reads held back
4. the test controller releases the reads

It probably needs to be more elaborate, since may succeed with the current code based on memory limiting. To demonstrate oom-kill we need memory consumption to increase post-query. Perhaps the delay for the index read can serve to trick r-c-s into admitting many queries, then the hammer falls with the unexpectedly large data read.

Reply:
Not sure it is worth just to make this work in debug. Release/dev should be enough to cover the logic.

Reply:
I don't propose introducing the virtualized files just for this.